pennylane/gradients/general_shift_rules.py

```diff
@@ -451,16 +402,26 @@ def generate_multishifted_tapes(tape, indices, shifts, multipliers=None):
             and ``multipliers``.
     """
     params = list(tape.get_parameters())
-    multipliers = np.ones_like(shifts) if multipliers is None else multipliers
+    shifts = qml.math.array(shifts)
+    if len(shifts.shape) == 1:
+        shifts = shifts.reshape((shifts.shape[0], 1))
+        indices = [indices]
+    if multipliers is None:
+        multipliers = np.ones_like(shifts)
+    else:
+        multipliers = qml.math.array(multipliers)
+        if len(multipliers.shape) == 1:
+            multipliers = multipliers.reshape((multipliers.shape[0], 1))
+
     tapes = []
 
-    for shift, multiplier in zip(shifts, multipliers):
+    for _shifts, _multipliers in zip(shifts, multipliers):
         new_params = params.copy()
         shifted_tape = tape.copy(copy_operations=True)
-        for idx, s, m in zip(indices, shift, multiplier):
-            dtype = new_params[idx].dtype
-            new_params[idx] = new_params[idx] * qml.math.convert_like(m, new_params[idx])
-            new_params[idx] = new_params[idx] + qml.math.convert_like(s, new_params[idx])
+        for idx, shift, multiplier in zip(indices, _shifts, _multipliers):
+            dtype = getattr(new_params[idx], "dtype", None) or type(new_params[idx])

```

Comment:
How come this change was needed here?

Reply:
This enables us to support standard `float` values. I thought that would be a nice feature, if users want to use parameter shift logic from the `gradients` module without employing an autodiff framework.
This  would happen for example in a bunch of tests of `gradients.finite_diff` (because the casting to the old dtype was not included in `generate_shifted_tapes` previously, this is a new issue.). A minimal example:
```python
dev = qml.device("default.qubit", wires=1)
@qml.qnode(dev)
def circuit(x, y):
    qml.RX(x, wires=0)
    qml.RY(y, wires=0)
    return qml.probs(0)

x = 0.6
y = 0.1
circuit(x, y)
circuit.qtape.trainable_params = {0, 1}
qml.gradients.finite_diff(circuit.qtape)
```
This would cause an error without the above change.

Reply:
Oh yes we definitely want to support this!

I guess I'm mostly confused about `type(new_params[idx])`. When would the `dtype` not exist, and what is `type()` returning in this case?

Reply:
Ah, for example in the case above, `x.dtype` does not exist (and therefore new_params[0] doesn't), because it is a simple python float. At the same time, `type(x)=float` :)

Reply:
Got it! Although it makes me think that we will _always_ want float here. To me, the logic seems to be:

- Get the autodiff framework float type, either `np.float32`, `np.float64`, `tf.float32`, etc.
- If the dtype doesn't exist, assume Python `float`

So it kinda sounds like we want something like this:

```python
dtype = getattr(new_params[idx], "dtype", float)
```

Reply:
I now went with this, assuming that all custom data types would be implemented in a way that defines `new_params[idx].dtype`.